로그인, 프로필 화면 [STEP1] liam777, wily

wilybear opened the pull request:
@wonhee009 
로그인, 프로필 화면 STEP1 PR입니다.

## 프로젝트 진행 방식
1시간씩 페어 프로그래밍 진행 후 같이 토론하며 리펙토링하였습니다.

## 주요 기능
- 로그인 화면 구현(스토리 보드)
- UITextField 입력 값 검증 및 버튼 활성화/비활성화
- Localization (Korean, English)

## 고민했던 점 및 궁금했던 점
1. 스토리보드 scrollview 사용
- dynamic type을 사용하면서 글자 크기가 커질 경우 화면이 잘릴 위험이 있어서 스크롤 뷰를 넣어서 해결했습니다. 이 방법이 일반적인 방법이 맞는지 궁굼합니다.

2. 이메일 입력 -> 검증 -> 버튼 활성화에 대한 기능 분리
- 이메일과 패쓰워드 텍스트 필드에 target-action 방식으로 값이 변경될 때 마다 각 이메일, 패스워드의 유효성을 검증하고 이를 변수 validateResults에 저장하여 관리하였습니다.
- 아이디, 패스워드에 대한 validate가 현재는 같은 로직이지만 실질적으로 달라진다고 생각해서 함수를 분리하여 작성하였습니다.
- validateResults 를 프로퍼티 옵저버를 통해 모두 true가 될 경우 로그인 버튼을 활성화하도록 하였습니다.

3. Custom Color사용
- 코드에서 Custom Color를 사용하기 위해서 Color+Extension 파일에서
```
static var buttonBackgroundColor: UIColor? = UIColor.init(named: "ButtonBackgroundColor")
```
이런식으로 선언해서 사용하였는데 다른 종은 방법이 존재하는지 궁금합니다.

wilybear committed:
feat: step1 로그인 화면 레이아웃 구현

- 스토리보드를 활용하여 로그인 화면 레이아웃 구현
- 커스텀 컬러 추가

liam-777 committed:
feat: step1 로그인 코드 구현

- 아이디, 비밀번호 유효성 검증
- Color+Extension 추가
- 키보드 처리

wilybear committed:
feat: Step1 Localization 추가

- Main Storyboard: Korean, English 지원
- ViewController: textFieldShouldReturn함수 코드 수정

wilybear committed:
feat: Step1 폰트 수정

- Dynamic type Font 수정

wonhee009 commented on LoginProfile/Color+Extension.swift:
```diff
@@ -0,0 +1,15 @@
+//
+//  Color+Extension.swift
```
`UIColor`에 대한 extension이니 파일명도 `UIColor+Extension`은 어떨까요?😀

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    @IBOutlet weak var loginButton: UIButton!
+    
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var pwTextField: UITextField!
```
`-TextField`로 네이밍한 점👍 어떤 UI 요소인지 명확한거 같아요.

`pwTextField`의 `pw`는 password의 약자로 보이는데요.
[공식 문서](https://www.swift.org/documentation/api-design-guidelines/#naming)에도 약어는 피하는게 좋다고 하네요☺️
<img width="739" alt="image" src="https://user-images.githubusercontent.com/30456206/148054260-fef7ac3a-6d3f-4add-a6ed-bea2a46f3e6a.png">

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        scrollView.addGestureRecognizer(touch)
+        emailTextField.delegate = self
+        pwTextField.delegate = self
+    }
```
`configure`에서 `setTitleColor`은 사용자 눈에 보여지는 부분을 설정하는 부분여서 다른 설정 부분들과 다른다고 느껴지는데요.
`func configureUI()`, `func configure()`처럼 화면에 보여지는 부분과 아닌 부분의 configure를 나눠도 좋을거 같네요:)

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        loginButton.setTitleColor(.labelDarkGrayColor, for: .disabled)
+        emailTextField.addTarget(self, action: #selector(emailTextFieldDidChange), for: .editingChanged)
+        pwTextField.addTarget(self, action: #selector(passwordTextFieldDidChange), for: .editingChanged)
+        let touch = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+        scrollView.addGestureRecognizer(touch)
```
키보드 이외의 부분을 tap하면 키보드가 내려가는 기능을 하기 위한 코드로 보이는데요.
이 부분도 `configure()`에 있는 것보다는 `func configureKeyboard`처럼 하나의 메서드로 빼는건 어떨까요?🤔

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        pwTextField.delegate = self
+    }
+    
+    func validateEmailInput(_ input: String?) -> Bool {
+        guard let input = input else { return false }
+        return input.count >= 5
+    }
```
Email인지 아닌지는 input의 문자열 길이만 보면 알 수 있을까요?🤔

liam-777 replied:
<img width="532" alt="스크린샷 2022-01-04 오후 9 28 39" src="https://user-images.githubusercontent.com/95397915/148059245-3b588f38-bfc2-4d73-a7a9-1b0d52210502.png">

이번 프로젝트 요구사항이 아이디, 패쓰워드에 대한 검증식 보다는 텍스트 필드 값 변화에 따라서 액션을 취하는 것으로 판단했습니다!

validate라는 네이밍이 오해를 산 것 같네요. 내일 wily랑 이야기해보고 수정 진행하겠습니다.

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    func validatePasswordInput(_ input: String?) -> Bool {
+        guard let input = input else { return false }
+        return input.count >= 5
+    }
```
`input.count >= 5`가 반복되고 있는데요.
5라는 수를 하나로 관리하는건 어떨까요?☺️

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var pwTextField: UITextField!
+    
+    var validateResults: [Bool] = [false, false] {
+        didSet(results) {
+            let isEnabled = results.reduce(true) { $0 && $1 }
+            self.toggleLoginButtonEnabled(isEnabled)
+        }
+    }
```
`validateResult`를 배열로 선언한 이유가 따로 있을까요?🤔
```swift
validateResults[0] = validateEmailInput(emailTextField.text)
```
이런식으로 직접 index에 접근해서 사용하고 있는데 만약 validate를 진행해야 하는 정보가 많다면 어떨까요?
어떤 index가 해당 정보에 대한 validate 결과값인지 바로 알아차리기 힘들거 같습니다🤔

wonhee009 reviewed (changes requested):
@wilybear @liam-777 첫 PR 축하합니다!🎉
UI가 있는 프로젝트는 처음이실텐데 어떠실까 궁금하네용!

코멘트 달아두었으니 확인하시고 이해 안되는 부분이나 더 얘기 나눠 보시고 싶은 부분은 언제든 DM 주세요!🙏

<img width="602" alt="image" src="https://user-images.githubusercontent.com/30456206/148043842-b893a291-ed20-4114-930c-00bba2efaa4f.png">
Dynamic Type에 대해서 화면 확인해보셨을까요?
현재 텍스트를 키우면 버튼에서부터 아래의 텍스트들이 잘려서 보이는데요. 의도된 바가 맞을까요?🤔

---

### 고민점

#### 1. 스토리보드 scrollview 사용

화면을 넘어가는 경우를 방지해야한다면 scrollview를 사용하는게 맞는거 같습니다👍

#### 2. 이메일 입력 -> 검증 -> 버튼 활성화에 대한 기능 분리

해당 코드에 코멘트 남겨두었습니다!😀

#### 3. Custom Color사용

작성해 주신 방법도 좋은 방법이라고 생각이 듭니다.
다만, 해당 컬러를 하나의 화면에서만 사용한다면 컬러를 해당 VC에 위치하는 것도 나쁘지 않을거 같습니다.
만약 저라면
```swift
extension UIColor {
    static func getColor(from name: String) -> UIColor? {
        UIColor.init(named: name)
    }
}
```
가져오는 메서드만 extension으로 구현하고 해당 화면에서 필요한 컬러를 뽑아다 쓸거 같긴합니다.

만약, 프로젝트 내에서 사용되는 컬러가 공통화 되어 있어서 하나의 컬러가 여러 화면에 대응된다면 현재 작성해주신 코드도 충분히 좋은 해결책이라고 생각합니다👍

wilybear committed:
feat: Step1 PR리뷰 반영

- Color+Extension -> UIColor+Extension
- ViewController.swfit의 viewdidload에서 configure 함수 분리
- validateResults -> isLoginButtonEnabled
- 변수 이름 리펙토링 진행

wonhee009 commented on LoginProfile/ViewController.swift:
```diff
@@ -11,39 +11,56 @@ class ViewController: UIViewController {
-        didSet(results) {
-            let isEnabled = results.reduce(true) { $0 && $1 }
-            self.toggleLoginButtonEnabled(isEnabled)
+    let minimumTextLength = 5
```
`private`여도 될거 같습니다! ☺️

wonhee009 reviewed (approved):
@wilybear @liam-777 수고하셨습니다!
코멘트 하나 남겨두었는데 step2에 반영해주시면 될거 같습니다!
